src/core/magnolia.scala

```diff
@@ -91,4 +92,138 @@ trait Derivation[TypeClass[_]] extends CommonDerivation[TypeClass]:
 end Derivation
 
 trait AutoDerivation[TypeClass[_]] extends Derivation[TypeClass]:
-  inline given autoDerived[A](using Mirror.Of[A]): TypeClass[A] = derived
\ No newline at end of file
+  inline given autoDerived[A](using Mirror.Of[A]): TypeClass[A] = derived
+
+trait MacroDerivation[TypeClass[_]]:
+  type Typeclass[T] = TypeClass[T]
+  def split[T](ctx: SealedTrait[Typeclass, T]): Typeclass[T]
+  def join[T](ctx: CaseClass[Typeclass, T]): Typeclass[T]
+
+  transparent inline def derived[T]: Typeclass[T] =
+    if isProduct[T] then
+      val cc = new CaseClass[Typeclass, T](
+        typeInfo[T],
+        isObject[T],
+        isValueClass[T],
+        MacroDerivation.getParams[Typeclass, T],
+        anns[T],
+        typeAnns[T]
+      ) {
+        def construct[PType](makeParam: Param => PType)(using ClassTag[PType]): T = ???
+        def rawConstruct(fieldValues: Seq[Any]): T = ???
+        def constructEither[Err, PType: ClassTag](makeParam: Param => Either[Err, PType]): Either[List[Err], T] = ???
+        def constructMonadic[M[_]: Monadic, PType: ClassTag](makeParam: Param => M[PType]): M[T] = ???
+      }
+      join(cc)
+    else if isSum[T] then
+      val sealedTrait = SealedTrait[Typeclass, T](
+        typeInfo[T],
+        MacroDerivation.getSubtypes[Typeclass, T],
+        List[Any](),
+        paramTypeAnns[T]
+      )
+      split(sealedTrait)
+    else
+      // summonInline[Typeclass[T]] // <-- this is tried to be summoned even if not needed
```

Comment:
probably `Expr.summon` wouldn't have this problem, as it's not inlined in the macro itself?